src/Sarif.Multitool/Rules/SARIF1012.MessageArgumentsMustBeConsistentWithRule.cs

```diff
@@ -0,0 +1,93 @@
+
+        /// <summary>
+        /// Placeholder
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF1012_MessageArgumentsMustBeConsistentWithRule_FullDescription_Text };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF1012_MessageArgumentsMustBeConsistentWithRule_Error_MessageIdMustExist_Text),
+            nameof(RuleResources.SARIF1012_MessageArgumentsMustBeConsistentWithRule_Error_SupplyEnoughMessageArguments_Text)
+        };
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        private IList<ReportingDescriptor> currentRules;
+        private Run run;
+
+        protected override void Analyze(Run run, string runPointer)
+        {
+            this.run = run;
+            this.currentRules = run.Tool.Driver?.Rules;
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            // if message.id is not null, we have to check if that exists in rules
+            if (!string.IsNullOrEmpty(result.Message.Id))
+            {
+                ReportingDescriptor rule = result.GetRule(this.run);
+
+                if (this.currentRules == null
+                    || rule.MessageStrings?.ContainsKey(result.Message.Id) == false)
+                {
+                    // {0}: Placeholder {1} {2}
+                    LogResult(
+                        resultPointer,
+                        nameof(RuleResources.SARIF1012_MessageArgumentsMustBeConsistentWithRule_Error_MessageIdMustExist_Text),
+                        result.Message.Id,
+                        result.ResolvedRuleId(run) ?? "null");
+                    return;
+                }
+
+                // we have rules, we find it but, the count of parameters are invalid
+                string messageText = rule.MessageStrings[result.Message.Id].Text;
+                int placeholderMaxPosition = PlaceholderMaxPosition(messageText);
+                if (placeholderMaxPosition > (result.Message.Arguments?.Count ?? 0))
+                {
+                    // {0}: Placeholder {1} {2} {3} {4} {5}
+                    LogResult(
+                        resultPointer,
+                        nameof(RuleResources.SARIF1012_MessageArgumentsMustBeConsistentWithRule_Error_SupplyEnoughMessageArguments_Text),
+                        result.Message.Arguments.Count.ToString(),
+                        result.Message.Id,
+                        result.ResolvedRuleId(run) ?? "null",
+                        placeholderMaxPosition.ToString(),
+                        messageText);
+                }
+            }
+        }
+
+        private int PlaceholderMaxPosition(string text)
+        {
+            MatchCollection matchCollection = Regex.Matches(text, "{\\d+}");
```

Comment:
>"{\\d+}" [](start = 66, length = 8)

You can use a regex "capture" to get just the integer part: `@"\{(<index>\d+)\}"`

Reply:
Also, `'{'` and `}` are regex metacharacters, so you must escape them (as I did above).

---
In reply to: [447722378](https://github.com/microsoft/sarif-sdk/pull/1944#discussion_r447722378) [](ancestors = 447722378)

Reply:
Finally, note that "verbatim strings" (`@"..."`) are very useful for regexes because you need lots of backslashes, and with a verbatim string you don't have to double the backslashes.

---
In reply to: [447723010](https://github.com/microsoft/sarif-sdk/pull/1944#discussion_r447723010) [](ancestors = 447723010,447722378)